eventually-v0-5-0/src/event.rs

```diff
@@ -0,0 +1,111 @@
+//! Module `event` contains types and abstractions helpful for working
+//! with Domain Events.
+
+use std::fmt::Debug;
+
+use async_trait::async_trait;
+use futures::stream::BoxStream;
+use serde::{Deserialize, Serialize};
+
+use crate::{
+    version::{ToConflictError, Version},
+    Message, Messages,
+};
+
+/// An Event is a [Message] carring the information about a Domain Event,
+/// an occurrence in the system lifetime that is relevant for the Domain
+/// that is being implemented.
+pub type Event<T> = Message<T>;
+
+/// Shortcut type to represent multiple [Event]s.
+pub type Events<T> = Messages<T>;
+
+/// An [Event] that has been persisted to the Event [Store].
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub struct Persisted<Id, Evt> {
+    /// The id of the Event Stream the persisted Event belongs to.
+    pub stream_id: Id,
+
+    /// The version of the Event Stream when this Event has been recorded.
+    ///
+    /// This value is used for optimistic concurrency checks, to avoid
+    /// data races in parallel command evaluations.
+    ///
+    /// Check the [Version] type and module documentation for more info.
+    pub version: Version,
+
+    /// The actual Domain Event carried by this envelope.
+    pub payload: Event<Evt>,
+}
+
+/// Shortcut type to represent multiple [Persisted] Events.
+pub type PersistedEvents<Id, Evt> = Vec<Persisted<Id, Evt>>;
+
+/// Specifies the slice of the Event Stream to select when calling [`Store::stream`].
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum VersionSelect {
+    /// Selects all [Events] in the Event [Stream].
+    All,
+
+    /// Selects all [Events] in the Event [Stream] starting from the [Event]
+    /// with the specified [Version].
+    From(Version),
+}
+
+/// Specifies an expectation on the Event [Stream] version targeted
+/// when calling [`Store::append`].
+///
+/// This type allows for optimistic concurrency checks, avoiding data races
+/// when modifying the same Event Stream at the same time.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum StreamVersionExpected {
+    /// Disables any kind of optimistic concurrency check, instructing the [Store]
+    /// to append the new [Event] no matter the current [Version] of the [Stream].
+    Any,
+
+    /// Sets the expectation that the Event [Stream] must be at the specified
+    /// [Version] for the [`Store::append`] call to succeed.
+    MustBe(Version),
+}
```

Comment:
this looks an awful lot like `Option<Version>`

Reply:
Same discussion about readability in https://github.com/get-eventually/eventually-rs/pull/192#discussion_r762906270 👀